app/frontend/src/features/messages/groupchats/AdminsDialog.tsx

```diff
@@ -0,0 +1,249 @@
+import { Error as GrpcError } from "grpc-web";
+import React, { useEffect, useState } from "react";
+import { useMutation, useQueryClient } from "react-query";
+
+import Alert from "../../../components/Alert";
+import Avatar from "../../../components/Avatar";
+import Button from "../../../components/Button";
+import ConfirmationDialogWrapper from "../../../components/ConfirmationDialogWrapper";
+import {
+  Dialog,
+  DialogActions,
+  DialogContent,
+  DialogTitle,
+} from "../../../components/Dialog";
+import IconButton from "../../../components/IconButton";
+import { AddIcon, CloseIcon } from "../../../components/Icons";
+import TextBody from "../../../components/TextBody";
+import { User } from "../../../pb/api_pb";
+import { GroupChat } from "../../../pb/conversations_pb";
+import { service } from "../../../service";
+import { useAuthContext } from "../../auth/AuthProvider";
+import useUsers from "../../userQueries/useUsers";
+import { useMembersDialogStyles } from "./MembersDialog";
+
+function AdminListItem({
+  groupChatId,
+  member,
+  memberIsAdmin,
+  setError,
+}: {
+  groupChatId: number;
+  member: User.AsObject;
+  memberIsAdmin: boolean;
+  setError: (value: string) => void;
+}) {
+  const classes = useMembersDialogStyles();
+
+  const isCurrentUser = useAuthContext().authState.userId === member.userId;
+
+  const queryClient = useQueryClient();
+  const clearError = () => setError("");
+  const handleError = (error: GrpcError) => setError(error.message);
+  const invalidate = () => {
+    queryClient.invalidateQueries(["groupChatMessages", groupChatId]);
+    queryClient.invalidateQueries(["groupChats"]);
+    queryClient.invalidateQueries(["groupChat", groupChatId]);
+  };
+
+  const makeAdmin = useMutation<Empty, GrpcError, void>(
+    () => service.conversations.makeGroupChatAdmin(groupChatId, member),
+    {
+      onMutate: clearError,
+      onSuccess: () => {
+        const previousGroupChat = queryClient.getQueryData<GroupChat.AsObject>([
+          "groupChat",
+          groupChatId,
+        ]);
+        const newAdminUserIdsList = previousGroupChat?.adminUserIdsList ?? [];
+        newAdminUserIdsList.push(member.userId);
+        queryClient.setQueryData(["groupChat", groupChatId], {
+          ...previousGroupChat,
+          adminUserIdsList: newAdminUserIdsList,
+        });
+        invalidate();
+      },
+      onError: handleError,
+    }
+  );
+  const removeAdmin = useMutation<Empty, GrpcError, void>(
+    () => service.conversations.removeGroupChatAdmin(groupChatId, member),
+    {
+      onMutate: clearError,
+      onSuccess: () => {
+        const previousGroupChat = queryClient.getQueryData<GroupChat.AsObject>([
+          "groupChat",
+          groupChatId,
+        ]);
+        const newAdminUserIdsList = previousGroupChat?.adminUserIdsList ?? [];
+        newAdminUserIdsList.splice(
+          newAdminUserIdsList.indexOf(member.userId),
+          1
+        );
```

Comment:
This seems pretty old fashioned but any other way seemed convoluted.

Reply:
Oh I was meant to do `Array.from` to copy it. I'd rather avoid mutating the original even if it works